docs/design/node_taint_eviction.md

```diff
@@ -0,0 +1,78 @@
+## General Design Goals
+
+The goal of this design is to implement a mechanism that allows the operator to replace Pods that are running on nodes with a taint.
+
+## Current Implementation
+
+We don't have a current implementation of this and the operator is currently not able to access any node information.
+The closest implementation to this is the [Replace Failed ProcessGroups reconciler](https://github.com/FoundationDB/fdb-kubernetes-operator/blob/main/controllers/replace_failed_process_groups.go), which could be extended for this.
+
+## Proposed Design
+
+We have to make sure that this feature can be enabled and disabled, e.g. if someone wants to run the operator inside a cluster without node access.
+The `AutomaticReplacementOptions` will be extended to support configuration of replacements for Pods that run on tainted nodes.
+The new field will be called `taints` and will contain a list, default would be empty, of taint configurations.
+That configuration will include the taint key, this has to be an exact match or a wildcard `*`, and the duration this taint must be present on the node.
+The `durationInSeconds` allow to specify how sensitive the operator should be to taints.
+If a wildcard is specified as key this will be used as default value, except there is an exact match defined for a taint key.
+
+```yaml
+...
+automaticReplacementOptions:
+  enabled: true
+  failureDetectionTimeSeconds: 7200
+  maxConcurrentReplacements: 1
+  enableNodeFailureDetection: true
+  taints:
+  - key: "example.org/maintenance"
+    durationInSeconds: 7200 # Ensure the taint is present for at least 2 hours before replacing Pods on a node with this taint
+  - key: "*" # The wildcard would allow to define a catch all configuration
+    durationInSeconds: 3600 # Ensure the taint is present for at least 1 hour before replacing Pods on a node with this taint
+```
+
+The actual implementation will extend the [ReplaceFailedProcessGroups](https://github.com/FoundationDB/fdb-kubernetes-operator/blob/main/internal/replacements/replace_failed_process_groups.go) to handle those cases.
+
+```go
+    if !cluster.NodeFailureDetectionEnabled() {
+		return hasReplacements
+    }
+
+	// Only replace process groups if we can do a replacement. We could also track this as an independent metric.
+	if maxReplacements <= 0 {
+		return hasReplacements
+    }
+
+	pods, err := r.PodLifecycleManager.GetPods(ctx, r, cluster, internal.GetPodListOptions(cluster, "", "")...)
+    if err != nil {
+        return &requeue{curError: err}
+    }
+
+	nodeMap := internal.GetNodeMap(pods) // This will generate a map containing the node name as key and the pods as value.
+	for node, pods := range nodeMap {
+		// Check if node has taint for the expected duration
+		// in that case we will replace all Pods of that node.
+    }
+```
+
+The limitation of this approach would be that we only detect Pods running on tainted nodes when a reconciliation is happening (those are not coupled to node events).
```

Comment:
we can use the periodic reconciliation in operator as a starting point? 
Since we don't do anything until a taint has been there for a certain duration (say 2 hours), periodically check should be timely enough.

Reply:
Yes, I thought I expressed that with this statement. Let me add some more context to make it clear we are targeting for the simple approach first.